distributed/diagnostics.py

```diff
@@ -0,0 +1,139 @@
+import logging
+import sys
+import threading
+import time
+from timeit import default_timer
+
+from .utils import ignoring
+from .executor import default_executor
+
+
+logger = logging.getLogger(__name__)
+
+
+class Diagnostic(object):
+    def start(self, scheduler):
+        pass
+
+    def task_finished(self, scheduler, key, worker, nbytes):
+        pass
+
+    def update_graph(self, scheduler, dsk, keys, restrictions):
+        pass
+
+
+class ProgressBar(Diagnostic):
+    def __init__(self, scheduler=None, keys=None, minimum=0, width=40, dt=0.1):
+        self._minimum = minimum
+        self._width = width
+        self._dt = dt
+        self.last_duration = 0
+
+        self.all_keys = set()
+        self.keys = set()
+
+        self._running = False
+
+        if scheduler is None and default_executor():
+            scheduler = default_executor().scheduler
+
+        if scheduler:
+            self.scheduler = scheduler
+            scheduler.add_diagnostic(self)
```

Comment:
The `scheduler is None and not default_executor()` case is unhandled. Not sure if that's important.